adapter/internal/discovery/xds/rate_limiter_cache.go

```diff
@@ -0,0 +1,265 @@
+	r.apiLevelMu.Lock()
+	defer r.apiLevelMu.Unlock()
+	delete(r.apiLevelRateLimitPolicies[org][vHost], apiID)
+}
+
+func getRateLimitPolicy(policies map[string]*mgw.APIRateLimitPolicy, policyName string) (*rls_config.RateLimitPolicy, error) {
+	policy, ok := policies[policyName]
+	if !ok {
+		return nil, fmt.Errorf("rate limit policy %q not defined", policyName)
+	}
+
+	unit, err := getRateLimitUnit(policy.SpanUnit)
+	if err != nil {
+		return nil, err
+	}
+
+	return &rls_config.RateLimitPolicy{
+		Unit:            unit,
+		RequestsPerUnit: policy.Count,
+	}, nil
+}
+
+func (r *rateLimitPolicyCache) generateRateLimitConfig(label string) *rls_config.RateLimitConfig {
+	var orgDescriptors []*rls_config.RateLimitDescriptor
+
+	r.apiLevelMu.RLock()
+	defer r.apiLevelMu.RUnlock()
+
+	// Generate API level rate limit configurations
+	for org, orgPolicies := range r.apiLevelRateLimitPolicies {
+		var vHostDescriptors []*rls_config.RateLimitDescriptor
+		for vHost, vHostPolicies := range orgPolicies {
+			var apiDescriptors []*rls_config.RateLimitDescriptor
+			for apiID, apiPolicies := range vHostPolicies {
+				// Configure API Level rate limit policies only if, the API is deployed to the gateway label
+				// Check API deployed to the gateway label
+				if arrayContains(orgIDOpenAPIEnvoyMap[org][apiID], label) {
+					apiDescriptors = append(apiDescriptors, apiPolicies...)
+				}
+			}
+			vHostDescriptor := &rls_config.RateLimitDescriptor{
+				Key:         envoyconf.DescriptorKeyForVhost,
+				Value:       vHost,
+				Descriptors: apiDescriptors,
+			}
+			vHostDescriptors = append(vHostDescriptors, vHostDescriptor)
+		}
+		orgDescriptor := &rls_config.RateLimitDescriptor{
+			Key:         envoyconf.DescriptorKeyForOrg,
+			Value:       org,
+			Descriptors: vHostDescriptors,
+		}
+		orgDescriptors = append(orgDescriptors, orgDescriptor)
+	}
+
+	return &rls_config.RateLimitConfig{
+		Name:        envoyconf.RateLimiterDomain,
+		Domain:      envoyconf.RateLimiterDomain,
+		Descriptors: orgDescriptors,
+	}
+}
+
+func (r *rateLimitPolicyCache) updateXdsCache(label string) bool {
+	rlsConf := r.generateRateLimitConfig(label)
+
+	version := fmt.Sprint(rand.Intn(maxRandomInt))
+	snap, err := gcp_cache.NewSnapshot(version, map[gcp_resource.Type][]gcp_types.Resource{
+		gcp_resource.RateLimitConfigType: {
+			rlsConf,
+		},
+	})
+	if err != nil {
+		loggers.LoggerXds.Error("Error while updating the rate limit snapshot: ", err)
+		return false
+	}
+	if err := snap.Consistent(); err != nil {
+		loggers.LoggerXds.Error("Inconsistent rate limiter snapshot: ", err)
+		return false
+	}
+
+	if err := r.xdsCache.SetSnapshot(context.Background(), label, snap); err != nil {
+		loggers.LoggerXds.Error("Error while updating the rate limit snapshot: ", err)
+		return false
+	}
+	loggers.LoggerXds.Infof("New rate limit cache updated for the label: %q version: %q", label, version)
+	loggers.LoggerXds.Trace("Updated rate limit config", rlsConf)
```

Comment:
Let's change this to Debug level as trace level is not supported from implementation. (Or else we could bring the trace level support as well)

Reply:
Changed to debug with https://github.com/wso2/product-microgateway/pull/3178/commits/55867ecbef7b3e452714f169f99fd86910881008